### PR TITLE
Adding tests for autofixing validate quotes

### PR DESCRIPTION
**Pull request by TheSavior**
Continuing off of #1201 and #1202. I don't have push permissions to that PR and I needed to rebase the branch, so sadly I had to create a new PR.

This is rebased off of #1264 to be able to have the new test fix assertions. That PR should be merged first.

The code seemed to pass the tests with no additional changes. The code currently does not fix multi line strings, but I think this is a great first step and is functionally mergeable as is. 

@indexzero Awesome job on this.

### CODE REVIEW

**Comment by TheSavior** [comment]
Also, I have run this on our large code base and didn't find any unintended consequences.

**Comment by mikesherov** [comment]
Can you move the one file to `test/lib/assertHelpers`?

**Comment by mikesherov** [comment]
This LGTM with that one suggested change. Thanks again as usual @TheSavior !

**Comment by TheSavior** [comment]
@mikesherov, that comment is applicable to #1264. I'll move it there and rebase this. That PR should be merged first.

**Comment by zxqfox** [comment]
@TheSavior #1264 was merged!

**Comment by TheSavior** [comment]
Alright, rebased off master.

**Comment by zxqfox** [comment]
Something goes wrong definitely:
- You adding test/assertHelpers.js in 0940a14
- Then you moving it out in 991a7a0

I'd try to put b3b3ff1 before the rest and squash the rest to one.

Like start `git rebase -i origin/master`, and put instructions like:
```
pick b3b3ff1
pick 0940a14
f 8e93382
f 991a7a0
```

Can you try it?

**Comment by TheSavior** [comment]
There we go. I just cleaned it up and squashed it all down to one.

**Comment by TheSavior** [comment]
*phew*. Builds passed.

**Comment by zxqfox** [comment]
:+1: Fine now ;-) Thanks!

LGTM.

@markelog @mikesherov Final look?

**Comment by TheSavior** [comment]
@zxqfox @mikesherov marked it as reviewed-once before the rebase.

**Comment by markelog** [comment]
Needed @mdevils to sign off, i'd say it should be done either through `token-assert` module or though `fix` property in rule declaration

**Comment by TheSavior** [comment]
For reference, the discussion about how to modify `errors` to enable this behavior started here: https://github.com/jscs-dev/node-jscs/pull/1202

File: lib/rules/validate-quote-marks.js [hunk 1]
```diff
@@ -106,10 +106,16 @@ module.exports.prototype = {
                     return;
                 }
 
+                token.value = token.value.replace(
+                    new RegExp('^' + mark + '(.*)' + mark + '$'),
```
Comment (markelog) [inline lib/rules/validate-quote-marks.js]: scarrrry

Comment (zxqfox) [inline lib/rules/validate-quote-marks.js]: Can't get how ruin this ;-) You know string without `'` or `"`? ES6 template strings?

Comment (TheSavior) [inline lib/rules/validate-quote-marks.js]: Yeah, came from #1202 that I pulled into this one. It seems to work and pass the tests, so I'm not super worried about it. It might need to change if we try to fix multi line strings, but I wouldn't bother changing it until that comes up. We shouldn't be scared of regex just because it is regex. Even though we both seemed to have the same reaction :P

**Comment by zxqfox** [comment]
@TheSavior Yeah, but: "Who knew life, that is no longer hurries" © Omar Khayyam

@markelog Thanks for summoning. ;-)

File: lib/rules/validate-quote-marks.js (hunk 1 above)
Comment (gibson042) [inline lib/rules/validate-quote-marks.js]: That seems overly complicated and slow for what it's doing (which is basically `token.value = opposite[mark] + token.value.slice(1, -1) + opposite[mark]`), and what it's doing seems inaccurate:
* as noted, it will fail on multi-line strings because `.` doesn't match newline (jQuery uses `[\w\W]` for a true "any character" match)
* it will also fail on strings containing unescaped quote characters, e.g. converting raw `"'"` into `'''`

I _**think**_ what you want to do is escape all unescaped quote characters _in_ the string, then change the quote characters _of_ the string... something like:
```js
var unescapedOpposite = {
	'"': /'/g,
	"'": /"/g
};
…
token.value = opposite[mark] +
	str.slice(1, -1).split(/((?:\\[\w\W])+)/).map(function(part, i) {
		return i % 2 === 0 ?
			part.replace(unescapedOpposite[mark], "\\$&") :
			part;
	}).join("") +
	opposite[mark];
```

It's also possible to do with a probably-faster but arguably more complex single-pass `replace`:
```js
var unescapedOpposite = {
	'"': /((?:\\[\w\W]|[^\\])*?)(')/g,
	"'": /((?:\\[\w\W]|[^\\])*?)(")/g
};
…
var start = 0;
token.value = opposite[mark] +
	token.value.slice(1, -1).replace(unescapedOpposite[mark], function(match, prefix, mark, offset) {
		var escaped = offset === start ?
			prefix + "\\" + mark :
			match;
		start = offset + match.length;
		return escaped;
	}) +
	opposite[mark];
```

Comment (zxqfox) [inline lib/rules/validate-quote-marks.js]:
```
str = '"asd qwe"';
re = new RegExp('^' + '"' + '(.*)' + '"' + '$');
function _swap(str, q) { return q + str.substr(1, str.length - 2) + q; }

str.replace(re, "'$1'"); // '\'asd qwe\''
_swap(str, '\'') // '\'asd qwe"\''

console.time(); for (var i = 0; i < 1000000; i++) str.replace(re, "'$1'"); console.timeEnd(); // 617ms
console.time(); for (var i = 0; i < 1000000; i++) _swap(str, '\''); console.timeEnd(); // 115ms
```

Safer and faster. Thoughts?

Comment (markelog) [inline lib/rules/validate-quote-marks.js]: I'd would use something time-tested, like https://github.com/sindresorhus/to-single-quotes/blob/master/index.js

Comment (zxqfox) [inline lib/rules/validate-quote-marks.js]: :+1:

Comment (TheSavior) [inline lib/rules/validate-quote-marks.js]: I'm accepting Pull Requests. ;) I didn't write the logic on this one, I pulled it from another PR and updated #1201 and #1202 here.

Comment (mdevils) [inline lib/rules/validate-quote-marks.js]: It will probably fail on this:

```
var string = "123'456"; // -> '123'456' — syntax error
```

And also. Let's not change token values inside rules. Let's add another assertion to Assert module.

**Comment by mikesherov** [comment]
@mdevils, there's no reusability for that assertion. I'd vote to keep it out of assertion library until we find reuse for it to keep assertion library from growing out of control.

**Comment by markelog** [comment]
@mdevils +1 to the Mike point

**Comment by mdevils** [comment]
@markelog @mikesherov OK

**Comment by markelog** [comment]
So we decided who not to do, who about we decide how to do it :-).

https://github.com/jscs-dev/node-jscs/issues/1113

?

**Comment by mdevils** [comment]
@markelog my parser failed to process your message)

**Comment by markelog** [comment]
We either fix stuff in `check` method or introduce `fix` property for this "specific" rules that could not be included in the `token-assert` and try to isolate its execution hence link to the relevant issue.
